metricflow/cli/main.py

```diff
@@ -721,5 +733,181 @@ def validate_configs(cfg: CLIContext, dw_timeout: Optional[int] = None, skip_dw:
+)
+@click.option(
+    "--solver-threshold",
+    type=click.FloatRange(min=0.5, max=1),
+    required=False,
+    default=0.6,
+    help="The inference solver weight threshold.",
+)
+@click.option(
+    "--solver-weights",
+    type=click_custom.ListParamType(
+        value_converter=lambda weight_str: int(weight_str),
+        min_length=4,
+        max_length=4,
+    ),
+    required=False,
+    default="1,2,4,8",
+    help="Comma-separated list of 4 integer weights to be assigned for each confidence value. "
+    "Interpreted weights will be assigned to confidences LOW,MEDIUM,HIGH,VERY_HIGH",
+)
+@click.option(
+    "--output-dir",
+    type=click.Path(
+        dir_okay=True,
+        file_okay=False,
+        writable=True,
+        readable=True,
+    ),
+    required=True,
+    help="Output directory for the inferred config files.",
+)
+@click.option(
+    "--overwrite",
+    is_flag=True,
+    required=True,
+    default=False,
+    help="If specified, allows existing configuration files to be overwritten by inference.",
+)
+@beta_feature_warning(feature_name="Data Source Inference")
+@pass_config
+@exception_handler
+@log_call(module_name=__name__, telemetry_reporter=_telemetry_reporter)
+def infer(
+    cfg: CLIContext,
+    tables: Optional[List[SqlTable]],
+    schema: Optional[str],
+    max_sample_size: int,
+    solver_threshold: float,
+    solver_weights: List[int],
+    output_dir: str,
+    overwrite: bool,
+) -> None:
+    """Infer data source configurations from warehouse information."""
+
+    if cfg.sql_client.sql_engine_attributes.sql_engine_type is not SupportedSqlEngine.SNOWFLAKE:
+        click.echo(
+            "Data Source Inference is currently only supported for Snowflake. "
+            "We will add support for all the other warehouses before it becomes a "
+            "stable feature. Stay tuned!"
+        )
+        return
+
+    if tables is None and schema is None:
+        raise click.UsageError("Either `--tables` or `--schema` have to be provided.")
+
+    click.echo("Running data source inference...")
+
+    start_ms = int(time.perf_counter() * 1000)
+
+    if schema is not None and tables is None:
+        with _get_spin_context_manager_factory(f"🔍 Fetching available tables for schema `{schema}`")():
+            # we know it's a Snowflake client, but `list_tables` is not in the `SqlClient` interface.
+            tables_strs: List[str] = cfg.sql_client.list_tables(schema)  # type: ignore
+            tables = [SqlTable(schema_name=schema, table_name=table_name.upper()) for table_name in tables_strs]
+
+        if len(tables) == 0:
+            click.echo("Schema has no tables.")
+            return
+
+    provider = SnowflakeInferenceContextProvider(client=cfg.sql_client, tables=tables, max_sample_size=max_sample_size)
+
+    # set up the solver
+    def solver_weighter_function(confidence: InferenceSignalConfidence) -> int:
```

Comment:
CURIOUS QUESTION (no need to make changes here): is it necessary to allow the user to put custom weights?

Reply:
People shouldn't really need to, but I wanted to make the CLI as flexible as possible. If we didn't add that option, they'd have to implement a custom script just for that, which is much more painful and inconvenient than just doing `--solver weights 1,2,3,4`.

Reply:
they might still need custom scripts if they want to tune/add/remove individual rules because that would be a mess to add to the CLI.